src/view.c

```diff
@@ -114,38 +115,50 @@ view_matches_query(struct view *view, struct view_query *query)
 
 	if (query->identifier) {
 		const char *identifier = view_get_string_prop(view, "app_id");
-		if (!(identifier && match_glob(query->identifier, identifier))) {
+		match = bool_to_tristate(
+			identifier && match_glob(query->identifier, identifier));
+		if (match == LAB_STATE_DISABLED) {
 			return false;
 		}
 	}
 
 	if (query->title) {
 		const char *title = view_get_string_prop(view, "title");
-		if (!(title && match_glob(query->title, title))) {
+		match = bool_to_tristate(title && match_glob(query->title, title));
+		if (match == LAB_STATE_DISABLED) {
 			return false;
 		}
 	}
```

Comment:
I don't quite see why we have to use `match` here (or in fact anywhere at all). As these are all optional `and` conditions, can't we just `return false` if its requested in the query but doesn't match? Something along this line
```c
static bool
query_str_fail(const char *condition, const char *value)
{
	if (!condition) {
		return false;
	}
	return !(value && match_glob(condition, value));
}

bool
view_matches_query(struct view *view, struct view_query *query)
{
	if (query_str_fail(query->identifier, view_get_string_prop(view, "app_id")) {
		return false;
	}
	if (query_str_fail(query->title, view_get_string_prop(view, "title")) {
		return false;
	}
	if (query->window_type != - 1 && !view_contains_window_type(view, query->window_type)) {
		return false;
	}
	[..]
	if (query->shaded && query->shaded != bool_to_tristate(view->shaded)) {
		return false;
	}
	[..]
}
```

Reply:
We need to distinguish between everything matching and nothing tested; a view never matches an empty query. In the earlier implementation, this was done with a separate `empty` Boolean that was tracked alongside a match variable. The recent change dropped `empty` by making `match` a tristate, but the tristate was never triggered in these early checks. Hence, a positive match in these early tests was counted as an empty query, and the match was failing.

Reply:
This raises the questions of why do we need to return `false` for an empty query in the first place. The query argument itself is optional but that is handled in `run_if_action()`.

And if we really need it, a dedicated `if (view_query_is_empty(query)) { return false; }` seems easier to read.

Reply:
I don't know why an empty query should fail to match, so I applied the principle of Chesterton's fence: without understanding the behavior, I was unwilling to remove it.